// active-record-js/dist/migrations/columns/Column.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const mysql = __importStar(require("mysql"));
class Column {
    constructor(name, type, options = null) {
        this.allowNull = true;
        this.comment = '';
        this.name = name;
        this.type = type;
        if (options && options.hasOwnProperty('allowNull')) {
            this.allowNull = options.allowNull;
        }
        if (options && options.hasOwnProperty('default')) {
            this.default = options.default;
        }
        if (options && options.hasOwnProperty('length')) {
            this.length = options.length;
        }
    }
    toString() {
        const expressions = [
            mysql.escapeId(this.name),
            this.type
        ];
        if (this.length) {
            expressions.push(`(${this.length.toString()})`);
        }
        if (!this.allowNull) {
            expressions.push('NOT');
        }
        expressions.push('NULL');
        if (this.default !== undefined) {
            expressions.push('DEFAULT', this.default);
        }
        return expressions.join(' ').trim();
    }
}
exports.default = Column;
