// active-record-js/dist/connection/Database.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const md5_1 = __importDefault(require("md5"));
const mysql = __importStar(require("mysql"));
class Database {
    constructor(options) {
        this.database = options.database;
        this.host = options.host;
        this.name = options.name;
        this.password = options.password;
        this.port = options.port;
        this.user = options.user;
        this.pool = mysql.createPool({
            database: this.database,
            host: this.host,
            password: this.password,
            port: this.port,
            user: this.user
        });
    }
    getName() {
        return this.name;
    }
    md5() {
        return md5_1.default(this.host + this.user + this.password + this.database);
    }
    async run(query) {
        return new Promise((resolve, reject) => {
            this.pool.getConnection((error, connection) => {
                if (error) {
                    connection.destroy();
                    reject(error);
                }
                connection.query(query, (error, results) => {
                    connection.destroy();
                    if (error) {
                        reject(error);
                    }
                    resolve(results);
                });
            });
        });
    }
}
exports.default = Database;
