// active-record-js/dist/connection/DB.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Database_1 = __importDefault(require("./Database"));
const Logger_1 = require("../libraries/Logger");
class DB {
    static async beginTransaction() {
        return await this.run('START TRANSACTION');
    }
    static async commit() {
        return await this.run('COMMIT');
    }
    static create(options) {
        const db = new Database_1.default(options);
        this.active = options.name;
        this.databases.set(options.name, db);
    }
    static enableQueryLog() {
        this.isQueryLogEnabled = true;
    }
    static getActiveDatabase() {
        return this.databases.get(this.active);
    }
    static getQueryLog() {
        return this.logger.getLogs();
    }
    static async run(query) {
        if (this.isQueryLogEnabled) {
            this.logger.log(query);
        }
        return await this.getActiveDatabase().run(query);
    }
    static async rollBack() {
        return await this.run('ROLLBACK');
    }
    static setActiveDatabase(name) {
        return this.databases.get(name);
    }
}
exports.default = DB;
DB.databases = new Map();
DB.isQueryLogEnabled = false;
DB.logger = new Logger_1.Logger();
