// active-record-js/dist/migrations/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ColumnType;
(function (ColumnType) {
    // Integers
    ColumnType["BIG_INTEGER"] = "BIGINT";
    ColumnType["INTEGER"] = "INT";
    ColumnType["SMALL_INTEGER"] = "SMALLINT";
    ColumnType["MEDIUM_INTEGER"] = "MEDIUMINT";
    ColumnType["TINY_INTEGER"] = "TINYINT";
    // Fixed-Point
    ColumnType["DECIMAL"] = "DECIMAL";
    ColumnType["NUMERIC"] = "NUMERIC";
    // Floating-Point
    ColumnType["DOUBLE"] = "DOUBLE";
    ColumnType["FLOAT"] = "FLOAT";
    // Bit
    ColumnType["BIT"] = "BIT";
    // Text
    ColumnType["BLOB"] = "BLOB";
    ColumnType["CHAR"] = "CHAR";
    ColumnType["LONG_BLOB"] = "LONGBLOB";
    ColumnType["LONG_TEXT"] = "LONGTEXT";
    ColumnType["MEDIUM_BLOB"] = "MEDIUMBLOB";
    ColumnType["MEDIUM_TEXT"] = "MEDIUMTEXT";
    ColumnType["TEXT"] = "TEXT";
    ColumnType["TINY_TEXT"] = "TINYTEXT";
    ColumnType["VARCHAR"] = "VARCHAR";
    // Choice
    ColumnType["ENUM"] = "ENUM";
    ColumnType["SET"] = "SET";
    // Date
    ColumnType["DATE"] = "DATE";
    ColumnType["DATETIME"] = "DATETIME";
    ColumnType["TIME"] = "TIME";
    ColumnType["TIMESTAMP"] = "TIMESTAMP";
    ColumnType["YEAR"] = "YEAR";
})(ColumnType = exports.ColumnType || (exports.ColumnType = {}));
