// active-record-js/dist/query/relations/Relation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const models_1 = __importDefault(require("../../model/stores/models"));
const Builder_1 = __importDefault(require("../Builder"));
class Relation {
    constructor(model, related) {
        this.model = model;
        this.related = related;
    }
    getForeignKeyValue() {
        return this.model.getAttribute(this.foreignKey);
    }
    getLocalTableName() {
        return models_1.default.getModel(this.model.constructor.name).table;
    }
    getLocalKeyValue() {
        return this.model.getAttribute(this.localKey);
    }
    getRelatedTableName() {
        const Model = models_1.default.getModel(this.related);
        let table = '';
        if (!Model.hasOwnProperty('table')) {
            table = Model.name.toLowerCase() + 's';
        }
        else {
            table = Model.table;
        }
        return table;
    }
    buildQuery() {
        return new Builder_1.default().setModel(this.related);
    }
}
exports.default = Relation;
